### PR TITLE
Extend Aeson module 

### PR DESCRIPTION
**Pull request by adamczykm**
Extending Aeson module to provide needed functionality for jsonwsp workings.

### CODE REVIEW

File: src/Aeson.purs [hunk 1]
```diff
@@ -112,24 +163,153 @@ getField' decoder obj str =
...
+
+else instance (Traversable t, DecodeAeson a, DecodeJson (t Json)) => DecodeAeson (t a) where
+  decodeAeson (Aeson { numberIndex, patchedJson: AesonPatchedJson pJson }) = do
+    jsons :: t _ <- map AesonPatchedJson <$> decodeJson pJson
+    for jsons (\patchedJson -> decodeAeson (Aeson { patchedJson, numberIndex }))
+
+class GDecodeAeson (row :: Row Type) (list :: RL.RowList Type) | list -> row where
+  gDecodeAeson :: forall proxy. FO.Object Aeson -> proxy list -> Either JsonDecodeError (Record row)
+
+instance gDecodeAesonNil :: GDecodeAeson () RL.Nil where
```
Comment (ngua) [inline src/Aeson.purs:281]: There's some inconsistency in the use of instance names in this module. I'd personally prefer if they were omitted

File: src/Aeson.purs [hunk 2]
```diff
@@ -23,74 +23,125 @@
...
+    , caseArray: map toStringifiedNumbersJson >>> fromArray
+    , caseObject: map toStringifiedNumbersJson >>> fromObject
+    }
 
-instance DecodeAeson Aeson where
-  decodeAeson = pure
+-- | Recodes Json to Aeson.
+--   NOTE. The operation is costly as its stringifies given Json
+--         and reparses resulting string as Aeson.
+jsonToAeson :: Json -> Aeson
```
Comment (ngua) [inline src/Aeson.purs:136]: Why do we need this? Just for testing?

File: src/Aeson.purs [hunk 3]
```diff
@@ -23,74 +23,125 @@
...
+  pure $ Aeson { numberIndex, patchedJson }
 
-instance DecodeAeson Number where
-  decodeAeson = decodeAesonViaJson
+-------- Json <-> Aeson --------
 
-instance DecodeAeson Json where
-  decodeAeson = decodeAesonViaJson
+-- | Replaces indexes in the Aeson's payload with stringified
+--   numbers from numberIndex.
```
Comment (ngua) [inline src/Aeson.purs:119]: Missing continuation pipes here and below

File: src/Aeson.purs [hunk 4]
```diff
@@ -112,24 +163,153 @@ getField' decoder obj str =
...
+  lookup :: Json -> String -> Maybe Json
+  lookup j lbl = caseJsonObject Nothing (FO.lookup lbl) j
+
+  mkAeson :: Json -> Aeson
+  mkAeson json = Aeson { numberIndex, patchedJson: AesonPatchedJson json }
+
 infix 7 getField as .:
 
--- TODO: add getFieldOptional if ever needed.
+type AesonCases a =
```
Comment (ngua) [inline src/Aeson.purs:181]: I think a doc comment might be nice for this type

**Review by ngua — COMMENTED** [review]
I'm OK with these changes since they make more sense in the context of #164. @klntsky I'd like your input as well on this one please

File: src/Aeson.purs [hunk 5]
```diff
@@ -112,24 +163,153 @@ getField' decoder obj str =
...
+  note (UnexpectedValue $ toStringifiedNumbersJson asn) $
+    mkAeson <$> (foldM lookup pjson keys :: Maybe Json)
+  where
+  lookup :: Json -> String -> Maybe Json
+  lookup j lbl = caseJsonObject Nothing (FO.lookup lbl) j
+
+  mkAeson :: Json -> Aeson
+  mkAeson json = Aeson { numberIndex, patchedJson: AesonPatchedJson json }
+
 infix 7 getField as .:
```
Comment (klntsky) [inline src/Aeson.purs:179]: I'd move this operator definition closer to the `getField` function

File: src/Aeson.purs [hunk 6]
```diff
@@ -112,24 +163,153 @@ getField' decoder obj str =
...
+
+  -- will never get index out of bounds
+  unsafeIndex :: forall (x :: Type). Array x -> Int -> x
+  unsafeIndex arr ix = unsafePartial $ Array.unsafeIndex arr ix
+
+  -- will never encounter non int number
+  coerceNumber :: Number -> Int
+  coerceNumber = round
+
+constAesonCases :: forall (a :: Type). (forall x. x -> a) -> AesonCases a
```
Comment (klntsky) [inline src/Aeson.purs:217]: `(forall x. x -> a)` is equivalent to `a` from the users point of view. Why not as for just `a`?

File: test/Aeson.purs [hunk 7]
```diff
@@ -60,3 +83,56 @@ suite = do
...
+  group "caseAeson" do
+    test "caseObject" $ liftEffect do
+      let asn = jsonToAeson $ encodeJson { a: 10 }
+      (caseMaybeAeson _ { caseObject = Just <<< flip getField "a" }) asn `shouldEqual` (Just $ Right 10)
+    test "caseArray" $ liftEffect do
+      let asn = jsonToAeson $ encodeJson [ 10 ]
+      (caseMaybeAeson _ { caseArray = map decodeAeson <<< head }) asn `shouldEqual` (Just $ Right 10)
+    test "caseNumber" $ liftEffect do
+      let asn = jsonToAeson $ encodeJson 20222202
+      (caseMaybeAeson _ { caseNumber = Just }) asn `shouldEqual` (Just "20222202")
```
Comment (klntsky) [inline test/Aeson.purs:128]: > Just "20222202"

Why it's a string, not a BigNum or Int?

File: src/Aeson.purs (hunk 5 above)
Comment (adamczykm) [reply src/Aeson.purs:179]: Will do

File: src/Aeson.purs (hunk 2 above)
Comment (adamczykm) [reply src/Aeson.purs:136]: I think I use it in the other PR for errors reports

File: src/Aeson.purs (hunk 1 above)
Comment (adamczykm) [reply src/Aeson.purs:281]: Will do

File: src/Aeson.purs (hunk 4 above)
Comment (adamczykm) [reply src/Aeson.purs:181]: Ok

**Review by ngua — APPROVED** [review]

File: src/Aeson.purs [hunk 8]
```diff
@@ -112,24 +165,156 @@ getField' decoder obj str =
...
+      Just object -> gDecodeAeson object (Proxy :: Proxy list)
+      Nothing -> Left $ TypeMismatch "Object"
+
+else instance (Traversable t, DecodeAeson a, DecodeJson (t Json)) => DecodeAeson (t a) where
+  decodeAeson (Aeson { numberIndex, patchedJson: AesonPatchedJson pJson }) = do
+    jsons :: t _ <- map AesonPatchedJson <$> decodeJson pJson
+    for jsons (\patchedJson -> decodeAeson (Aeson { patchedJson, numberIndex }))
+
+class GDecodeAeson (row :: Row Type) (list :: RL.RowList Type) | list -> row where
+  gDecodeAeson :: forall proxy. FO.Object Aeson -> proxy list -> Either JsonDecodeError (Record row)
```
Comment (vvtran) [inline src/Aeson.purs:284]: Minor aesthetic but there's `FO.Object` and `Object` in the module due to the import
```
import Foreign.Object (Object)
import Foreign.Object as FO
```
The import is fine, but Is it worth sticking to one in the codebase?

File: src/Aeson.purs [hunk 9]
```diff
@@ -112,24 +165,156 @@ getField' decoder obj str =
...
+getNestedAeson asn@(Aeson { numberIndex, patchedJson: AesonPatchedJson pjson }) keys =
+  note (UnexpectedValue $ toStringifiedNumbersJson asn) $
+    mkAeson <$> (foldM lookup pjson keys :: Maybe Json)
+  where
+  lookup :: Json -> String -> Maybe Json
+  lookup j lbl = caseJsonObject Nothing (FO.lookup lbl) j
 
--- TODO: add getFieldOptional if ever needed.
+  mkAeson :: Json -> Aeson
+  mkAeson json = Aeson { numberIndex, patchedJson: AesonPatchedJson json }
```
Comment (vvtran) [inline src/Aeson.purs:179]: There seems to be two `mkAeson` inside  `where`. Maybe move to one place?

File: src/Aeson.purs [hunk 10]
```diff
@@ -112,24 +165,156 @@ getField' decoder obj str =
...
+  caseNull
+  caseBoolean
+  (coerceNumber >>> unsafeIndex numberIndex >>> caseNumber)
+  caseString
+  (map mkAeson >>> caseArray)
+  (map mkAeson >>> caseObject)
+  pJson
+  where
+  mkAeson :: Json -> Aeson
+  mkAeson json = Aeson { patchedJson: AesonPatchedJson json, numberIndex }
```
Comment (vvtran) [inline src/Aeson.purs:209]: See comment above.

File: src/Aeson.purs [hunk 11]
```diff
@@ -112,24 +165,156 @@ getField' decoder obj str =
...
+
+-------- Decode helpers --------
 
 -- | Ignore numeric index and reuse Argonaut decoder.
 decodeAesonViaJson
   :: forall (a :: Type). DecodeJson a => Aeson -> Either JsonDecodeError a
-decodeAesonViaJson (Aeson { json }) = decodeJson json
+decodeAesonViaJson (Aeson { patchedJson: AesonPatchedJson j }) = decodeJson j
 
 decodeAesonString
```
Comment (vvtran) [inline src/Aeson.purs:237]: Doc string here could be helpful

**Review by vvtran — COMMENTED** [review]
Few comments, otherwise looks good.

File: src/Aeson.purs [hunk 12]
```diff
@@ -112,24 +165,156 @@ getField' decoder obj str =
...
+
+      Nothing ->
+        Left $ AtKey fieldName MissingValue
+
+class DecodeAesonField a where
+  decodeAesonField :: Maybe Aeson -> Maybe (Either JsonDecodeError a)
+
+instance DecodeAeson a => DecodeAesonField (Maybe a) where
+  decodeAesonField Nothing = Just $ Right Nothing
+  decodeAesonField (Just j) = Just $ decodeAeson j
```
Comment (vvtran) [inline src/Aeson.purs:317]: We could write this as `decodeAesonField = Just <<< maybe (Right Nothing) decodeAeson` although there may be a preference for more explicit class instances to aid readability.

**Pull request merged**